core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -615,7 +615,11 @@ private[spark] class ExecutorAllocationManager(
       val taskIndex = taskEnd.taskInfo.index
       val stageId = taskEnd.stageId
       allocationManager.synchronized {
-        numRunningTasks -= 1
+        if (numRunningTasks > 0) {
+          numRunningTasks -= 1
```

Comment:
a better way to guard against duplicates is to keep track of a set of `runningTasks` instead.